---
title: "palmid development notebook"
output: html_notebook
---

## Initialization

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{r, include=FALSE, warning=FALSE}
# Development libraries
#library(roxygen2)
roxygen2::roxygenise()
  load("data/palmdb.Rdata")

#library(palmid)
```

```{r}
# INITIALIZE PALMID WORKSPACE -------------------
input.fev     <- 'data/waxsys.fev'
input.pro     <- 'data/waxsys.pro'

output.report <- 'data/waxsys_pp.png'
output.pro    <- 'data/waxsys_pro.png'
output.tax    <- 'data/waxsys_tax.png'
output.geo    <- 'data/waxsys_geo.html'
output.orgn   <- 'data/waxsys_orgn.png'

# ID Threshold for inclusion of a palmprint-match
# into the SRA workflow
id_threshold <- 0 #include all

# Import a palmprint-analysis
pp.in <- read.fev(input.fev, FIRST = TRUE)

# Import a diamond-alignd pro file
pro.in <- read.pro(input.pro)

# Establish Serratus server connection
con <- SerratusConnect()

```

## Input Palmprint Analysis

```{r}
# ANALYZE PALMPRINT -----------------------------
# Print out standard palmscan report
system('cat data/waxsys.txt')
```

```{r}
# Generate a palmid-report
pp.report <- PlotReport(pp.in, palmdb)
```

```{r, fig.height=6, fig.width=8, fig.retina=2}
# Print out palmid-report graphic
plot(pp.report)

# SAVE PP-REPORT --------------------------------
# png(filename = output.report, width = 800, height = 400)
#  plot(pp.report)
# dev.off()
```

## Palmprint vs. PalmDB Analysis

```{r}
# Print out MSA
system('cat data/waxsys.msa.fa')
```

```{r}
# ANALYZE PALMDB ALIGNMENT -----------------------------
pro.df     <- read.pro("data/waxsys.pro")
pro.report <- PlotProReport(pro.df, html = T)
```

```{r, fig.height=4, fig.width=8, fig.retina=2}
# Print protein-alignment report
ggplotly(pro.report)

# SAVE PP-REPORT 
# png(filename = output.pro, width = 800, height = 400)
#   plot(pro.report)
# dev.off()
```

```{r}
# Populate pro.df with PalmDB Taxonomic data, generate report
pro.df     <- get.proTax(pro.df, con)
tax.report <- PlotTaxReport(pro.df)
```

```{r, fig.height=10, fig.width=16, retina = 1}
# Create PalmDB Taxonomy plots
plot(tax.report)

# ggsave(output.tax,
#   plot(tax.report),
#   width = 16,
#   height = 10,
#   dpi = 72
# )
```

## PalmDB / SRA-Run Analysis

```{r}
# Perform Serratus SQL queries to retrieve
# parent/child sOTU lookup, sra, biosample, date, organism, geo
palm.sra <- get.palmSra(pro.df, con)
```

```{r}
# # For each parent and child sOTU, retrieve matching SRA runs
# ppdb.hits <- pro.df$sseqid[(pro.df$pident >= id_threshold)]
#   palm.uid  <- get.sOTU(ppdb.hits, con, get_childs = T)
#   palm.usra <- get.sra(palm.uid, con)
#   #palm.contigs<- get.sra(palm.group, con, get_contig.df = T)

# ANALYZE SRA/BIOSAMPLE SOURCES ---------------------------
geo.report <- PlotGeo2(palm.sra)

```

```{r, out.width = '100%'}
# Print geospatical-time report
geo.report

# SAVE GEO-REPORT 
#htmlwidgets::saveWidget(geo.report, file=output.geo)
```


```{r}
# Create Organism-tag wordcloud
orgn.report <- PlotOrgn(palm.sra)
```

```{r, fig.width=8, fig.height=5, retina = 2}
# Plot SRA-Organism Wordcloud
plot(orgn.report)

# # SAVE ORGN-REPORT 
# png(filename = output.orgn, width = 800, height = 400, res = 100)
#   plot(orgn.report)
# dev.off()

```
