---
title: "palmID: viral-RdRP report"
output:
  html_notebook:
    code_folding: hide
params:
  input.path: 'data/waxsys'
---

`palmID` is a contained analysis suite for viral RNA-dependent RNA polymerases
(RdRP) based on [Babaian and Edgar, 2021](https://www.biorxiv.org/content/10.1101/2021.03.02.433648v1).

## palmID Initialize

```{r setup, include=FALSE}
# RMarkdown Setting Initialization
# Command Line Interface ----

# Rscript -e "rmarkdown::render( \
#   input = 'palmid_dev.Rmd', \
#   output_file = 'amexnv.html', \
#   output_format = 'html_notebook', \
#   params=list( input.path = 'amexnv/amexnv'))"

# ---------------------------

knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{css, echo=FALSE}
/* Move code folding buttons to the left */
div.col-md-12 .pull-right {
  float: left !important
}
```

```{r, include=FALSE, warning=FALSE}
# Development libraries
production.version = TRUE

if (production.version){
  # Load stable palmid package
  library('palmid', quietly = T)
  load("data/palmdb.Rdata") 
  ### REQUIRES CONTAINER/RMD IS RUN FROM palmid/ REPO FOLDER
} else {
  
  # Compile a new palmid from source
  library(roxygen2)
    roxygen2::roxygenise()
    load("data/palmdb.Rdata")
}

```

```{r}
# INITIALIZE PALMID WORKSPACE -------------------
library('palmid')

# Establish Serratus server connection
con <- SerratusConnect()

# Input file
# Generated in `palmid` container
# params are defined in YAML header to expose to CLI
input.path      <- params$input.path

if (is.null(input.path)){
  stop("Error: No input provided.")
}

  input.pp      <- paste0(input.path, '.trim.fa')    # palmscan-palmprint sequence
  input.fev     <- paste0(input.path, '.fev')    # palmscan .fev report
  input.rep     <- paste0(input.path, '.txt')    # palmscan text motif-report
  input.pro     <- paste0(input.path, '.pro')    # diamond palmDB-alignment file
  input.msa     <- paste0(input.path, '.msa.fa') # muscle msa

# Output HTML Report
output.html   <- paste0(input.path, '.html')
save.plots    <- F # save individual plots as png

# Parameters
  id_threshold <- 0 # Minimum AA% to retain a hit

run.time <- Sys.time()

# IMPORT DATASETS -------------------------------
# Import a palmprint-analysis
pp.in <- read.fev(input.fev, FIRST = TRUE)

# Import a diamond-aligned pro file
pro.df <- read.pro(input.pro)
  # Populate with Taxonomy-data
  pro.df     <- get.proTax(pro.df, con)

# SQL-Import of palmprint/sra meta-data
  # parent/child sOTU lookup, sra, biosample, date, organism, geo
palm.sra <- get.palmSra(pro.df, con)

# GENERATE REPORT-PLOTS -------------------------
# Palmprint Report
pp.report <- PlotReport(pp.in, palmdb)

# Diamond-palmDB Alignment Report
pro.report <- PlotProReport(pro.df, html = T)

# PalmDB Viral Taxonomy Report
tax.report <- PlotTaxReport(pro.df)

# Geospatial distribution Report
geo.report <- PlotGeo2(palm.sra)

# Host/Library organism Report
orgn.report.freq <- PlotOrgn(palm.sra, freq = T)
orgn.report.pidn <- PlotOrgn(palm.sra, freq = F)

```


```{r, echo = FALSE}
pro.df %>%
  downloadthis::download_this(
    output_name = paste0(input.path, '.pro'),
    output_extension = ".csv",
    button_label = "Download palmDB.csv",
    button_type = "success",
    has_icon = TRUE,
    icon = "fa fa-file-export"
  )

palm.sra %>%
  downloadthis::download_this(
    output_name = paste0(input.path, '.sra.metadata'),
    output_extension = ".csv",
    button_label = "Download sra.metadata.csv",
    button_type = "success",
    has_icon = TRUE,
    icon = "fa fa-file-export"
  )
```

## RdRP palmprint QC

-- palmscan:

```{r}
# Palmscan-generated reports
system(paste0( 'echo "palmprint sequence: \n" ;
                cat ', input.pp, ';',
               'echo "\ncatalytic-motifs: " ;
                cat ', input.rep ))

#lapply(scan(input.pp, 'raw'),cat)[[1]]

```

-- palm-report:

```{r, fig.height = 4.5, fig.width = 9}
# Palmscan QC-plot
plot(pp.report)
```

## Palmprint vs. PalmDB Analysis

-- Multiple-Sequence Alignment (top 10 palmDB)

```{r}
# Multiple Sequence Alignment of Top-10 hits in palmDB
system( paste0 ('cat ', input.msa ))
```

-- Input palmprint aligned to palmDB

```{r, fig.height = 5, fig.width = 9}
# Protein-alignment of Input vs. palmDB
ggplotly(pro.report)

```
```{r, echo = FALSE}
# 
# blast.col <- linkBLAST(pro.df$sseqid, pro.df$full_sseq)
# 
# cbind( pro.df[, c("qseqid", "sseqid", 'tspe', "pident", "evalue", "cigar")],
#        blast.col) %>%
#   DT::datatable(
#     rownames = FALSE, filter = "top", escape = F,
#     options = list(pageLength = 10, scrollX = T)
#   )
# 

```

-- palmDB-hits Viral Taxonomy

```{r, fig.height=10, fig.width=16, retina = 1}
# Viral Taxonomy of palmDB Hits
plot(tax.report)
```

## SRA Meta-data Analysis

-- Geospatial distribution of palmprint-containing sequencing libraries
   (click to explore)

```{r, fig.height = 6, fig.width = 9}
# Geo-distribution of palmprint-containing SRA
geo.report
```

-- Source organism of sequencing libraries

```{r, fig.width=8, fig.height=5, retina = 2, out.width=c('50%', '50%')}
# Plot SRA-Organism Wordcloud - id
# (scaled by AA% proximity to input palmprint)
plot(orgn.report.pidn)
```

```{r, fig.width=8, fig.height=5, retina = 2}
# Plot SRA-Organism Wordcloud - freq
# (scaled by frequency in the SRA)
plot(orgn.report.freq)
```


#### Footer
- [Source code](https://github.com/ababaian/palmid)
- [Submit issue / ask question](https://github.com/ababaian/palmid/issues)

```{r, include=FALSE}
if (save.plots){
  
  # SAVE PP-REPORT
  png(filename = output.report, width = 800, height = 400)
   plot(pp.report)
  dev.off()
  
  # SAVE PRO-REPORT 
  png(filename = output.pro, width = 800, height = 400)
    plot(pro.report)
  dev.off()
  
  # SAVE TAX-REPORT
  ggsave(output.tax,
    plot(tax.report),
    width = 16,
    height = 10,
    dpi = 72
  )
  
  # SAVE GEO-REPORT
  htmlwidgets::saveWidget(geo.report, file=output.geo)

  # SAVE ORGN-REPORT
  png(filename = output.orgn, width = 800, height = 400, res = 100)
    plot(orgn.report)
  dev.off()

}

```

```{r, include = FALSE}
#rmarkdown::render('palmid_dev.Rmd',
#                  output_file = output.html)
```
