---
title: "palmID: RNA Virus RdRP Report"
output:
  html_notebook:
    code_folding: hide
params:
  input.path: 'inst/extdata/waxsys'
  prod.run: FALSE
---

`palmID` is a contained analysis suite for viral RNA-dependent RNA polymerases
(RdRP) based on the "Palmprint" RNA virus barcodes described by [Babaian and Edgar, 2021](https://www.biorxiv.org/content/10.1101/2021.03.02.433648v1).

## Initialize

```{r setup, include=FALSE}
# RMarkdown Setting Initialization
# Command Line Interface ----

# Rscript -e "rmarkdown::render( \
#   input = 'palmid_dev.Rmd', \
#   output_file = 'amexnv.html', \
#   output_format = 'html_notebook', \
#   params=list( input.path = 'amexnv/amexnv'))"

# ---------------------------

knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{css, echo=FALSE}
/* Move code folding buttons to the left */
div.col-md-12 .pull-right {
  float: left !important
}
```

```{r, include=FALSE, warning=FALSE}
# Run as Production Pipeline or Development
production.version = params$prod.run

if (production.version) {
  # Load stable palmid package
  library('palmid', quietly = T)
  data("palmdb") 

  # Disable warnings
   defaultW <- getOption("warn")
   options(warn = -1)
} else {
  # Warnings variable
  defaultW <- getOption("warn")
  options(warn = -1)

  # Compile a new palmid from source
  roxygen2::roxygenise()
  load("data/palmdb.RData")
}

```


```{r}
# INITIALIZE PALMID WORKSPACE -------------------
# library("palmid")

# Establish Serratus server connection
con <- SerratusConnect()

# Input file
# Generated within `palmid` container
# params are defined in YAML header to expose to CLI
input.path      <- params$input.path

if (is.null(input.path)) {
  stop("Error: No input provided.")
}

  input.fa      <- paste0(input.path, '.input.fa')    # palmscan-palmprint sequence
  input.pp      <- paste0(input.path, '.trim.fa')    # palmscan-palmprint sequence
  input.fev     <- paste0(input.path, '.fev')    # palmscan .fev report
  input.rep     <- paste0(input.path, '.txt')    # palmscan text motif-report
  input.pro     <- paste0(input.path, '.pro')    # diamond palmDB-alignment file
  input.msa     <- paste0(input.path, '.msa.fa') # muscle msa

# Output HTML Report
output.html   <- paste0(input.path, '.html')
save.plots    <- FALSE # save individual plots as png

# Parameters
  id_threshold    <- 0   # Minimum AA% to retain a hit
  max_palmdb_hits <- 500 # Maximum number of alignment hits in PalmDB hits to return

run.time <- Sys.time()

# IMPORT DATASETS -------------------------------
# Import a palmprint-analysis
pp.in <- read.fev(input.fev, FIRST = TRUE)

# Import a diamond-aligned pro file
pro.df <- read.pro(input.pro)
  # Populate with Nickname/Taxonomy-data
  pro.df$nickname     <- get.nickname(pro.df$sseqid, con, ordinal = T)
  pro.df     <- get.proTax(pro.df, con)
  
# Set backstop when too many similar hits come up
palmdb.hits <- length(pro.df$qseqid)
if (palmdb.hits > max_palmdb_hits) {
  pro.df <- pro.df[ 1:max_palmdb_hits, ]
  print.hitn <- paste0("Reporting Top ", max_palmdb_hits, "/", palmdb.hits, " matches")
} else {
  print.hitn <-  paste0("Reporting all ", palmdb.hits, " matches")
}

# SQL-Import of palmprint/sra meta-data
  # parent/child sOTU lookup, sra, biosample, date, organism, geo
palm.sra <- get.palmSra(pro.df, con)

# SQL-import of STAT kmer taxonomic analysis of the retrieved
# matching SRA libraries
stat.sra <- get.sraSTAT(palm.sra$run_id, con)
# Populate stat.sra with percent identity from palm.sra
  stat.sra$pident <- palm.sra$pident[ match(stat.sra$run_id, palm.sra$run_id) ]
  

# GENERATE REPORT-PLOTS -------------------------
# Palmprint Report
pp.report <- PlotReport(pp.in, palmdb)

# Diamond-palmDB Alignment Report
pro.report <- PlotProReport(pro.df, html = T)

# PalmDB Viral Taxonomy Report
tax.report <- PlotTaxReport(pro.df)

# Geospatial distribution Report
geo.report <- PlotGeo2(palm.sra)

# Host/Library organism Report
orgn.report <- PlotOrgn(palm.sra, freq = FALSE)
stat.report <- PlotSTAT(stat.sra)

```

#### Download Data

```{r, echo = FALSE}
options(warn = -1)

pro.df %>%
  downloadthis::download_this(
    output_name = paste0(input.path, '.pro'),
    output_extension = ".csv",
    button_label = "Download palmdb_alignment.csv",
    button_type = "success",
    has_icon = TRUE,
    icon = "fa fa-file-export"
  )

palm.sra %>%
  downloadthis::download_this(
    output_name = paste0(input.path, '.sra.metadata'),
    output_extension = ".csv",
    button_label = "Download sra_metadata.csv",
    button_type = "success",
    has_icon = TRUE,
    icon = "fa fa-file-export"
  )
```

#### Summary

```{r, echo = FALSE}
# Generate text-based summary of palmID data
palm.summary <- text.summary(pp.in, pro.df, palm.sra, palmdb.hits)
cat(palm.summary)
```

## Input RdRP palmprint QC

#### palmscan motif analysis

Identification of the the core catalytic motifs A,B,C 
within the input sequence and reporting the "palmprint"
RNA-virus barcode.

```{r, echo = FALSE}
# Input Sequence
cat("Input FASTA sequence: \n\n")
cat(paste0(readLines(input.fa), collapse = "\n"), "\n")
cat("\n")

# Palmscan-generated reports
cat("palmprint sequence: \n\n")
cat(paste0(readLines(input.pp), collapse = "\n"), "\n")
cat("\n")

cat("catalytic-motifs: \n")
cat(paste0(readLines(input.rep), collapse = "\n"), "\n")
cat("\n")

# lapply(scan(input.rep, 'character'),cat)[[1]]

```

#### palmprint QC-report

Quality control metrics on the motif-scores and size distribution
of the "palmprint" informs the confidence that the input sequence
contains a viral RdRP. Input scores are shown compared to score 
distribution from 15,010 canonical viral RdRP from GenBank.

```{r, fig.height = 4.5, fig.width = 9}
# Palmscan QC-plot
plot(pp.report)
```

## Input Palmprint vs. PalmDB Analysis

#### Multiple Sequence Alignment

A `MUSCLE`-generated multiple sequence alignment of the 
top-10 matching hits in [`palmdb`](https://github.com/rcedgar/palmdb).

```{r}
# Multiple Sequence Alignment of Top-10 hits in palmDB
cat(paste0(readLines(input.msa), collapse = "\n"))
```

#### Input palmprint alignment to palmDB

Input "palmprint" are aligned against [`palmdb`](https://github.com/rcedgar/palmdb)
using `diamond` to retrieve related viruses. Upto the top 500 related viruses
are reported.


```{r, fig.height = 5, fig.width = 9}
# Protein-alignment of Input vs. palmDB
# Taxonomic Demarcations (~)
# Species >90%   - Red
# Genus   70-90% - Orange
# Family  45-70% - Green
# Phylum  <45%   - Purple

plotly::ggplotly(pro.report)

```
![](img/tax_legend.png)

#### Virus Taxonomy of PalmDB matches

Species-, Genus- and Phylum- level taxonomy of the matching
hits in PalmDB allow for taxonomic-inference of the input
virus.

```{r, fig.height=10, fig.width=16, retina = 1}
# Viral Taxonomy of palmDB Hits
plot(tax.report)
```
#### palmDB Alignment table

```{r, echo = FALSE}
### Data-Table: `palmDB.csv`
# Data table for palmDB Alignment

blast.col <- linkBLAST(pro.df$sseqid, pro.df$full_sseq)

cbind( pro.df[, c("sseqid", 'tspe', "pident", "evalue", "cigar")],
       blast.col) %>%
  DT::datatable(
    rownames = FALSE, filter = "top", escape = FALSE,
    options = list(pageLength = 10, scrollX = TRUE)
  )

```
[Download palmDB Alignment Table](#download-data)

## SRA Meta-data Analysis

RNA viruses matching input are cross-referenced against
5.7M SRA sequencing libraries to identify sra-virus matches
and their associated meta-data.

#### Geospatial distribution of related RNA viruses

Distribution of matching RNA viruses. Click to explore
each matching record. Species, genus, family and phylum
level matches are red, orange, green and purple, respectively.

```{r, fig.height = 6, fig.width = 9}
# Geo-distribution of palmprint-containing SRA
# Taxonomic Demarcations (~)
# Species >90%   - Red
# Genus   70-90% - Orange
# Family  45-70% - Green
# Phylum  <45%   - Purple

geo.report

```
![](img/tax_legend.png)

#### Virus-Organism Associations

Organism meta-data from the SRA sequencing libraries matching the input-virus.
Word size and color is scaled by proximity of the input-virus to its match in 
the library. Wordclouds show a) Researcher provided annotations of the libraries
or b) STAT automated taxonomic (orders) analysis of the reads contained within
the library ([Katz et al., 2021](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-021-02490-0)).

```{r, fig.width=8, fig.height=5, retina = 2}
# Plot SRA  Wordcloud - id
# (scaled by AA% proximity to input palmprint)
plot(orgn.report)
```

```{r, fig.width=8, fig.height=5, retina = 2}
# Plot SRA-Organism Wordcloud - freq
# (scaled by frequency in the SRA)
plot(stat.report)
```
![](img/stat_legend.png)

#### Input Alignment to SRA-RdRp Table

```{r, echo = FALSE}
### Data-Table: `sra.metadata.csv`
# Data table for palmDB Alignment

blast.col <- linkBLAST(palm.sra$sra_sequence,
                       paste0(palm.sra$run_id,"_",palm.sra$palm_id)
                       )

cbind( palm.sra[, c("run_id", "biosample_id", 'palm_id', 'pident', 'coverage', 'evalue', 'scientific_name')],
       blast.col) %>%
  DT::datatable(
    rownames = FALSE, filter = "top", escape = F,
    options = list(pageLength = 10, scrollX = T)
  )

```

[Download SRA Metadata Table](#download-data)

#### Footer

- [Source code](https://github.com/ababaian/palmid)
- [Submit issue / ask question](https://github.com/ababaian/palmid/issues)

To save this report: `File` --> `Save Page As`.

```{r, include=FALSE}
if (save.plots) {
  
  # SAVE PP-REPORT
  png(filename = output.report, width = 800, height = 400)
   plot(pp.report)
  dev.off()
  
  # SAVE PRO-REPORT 
  png(filename = output.pro, width = 800, height = 400)
    plot(pro.report)
  dev.off()
  
  # SAVE TAX-REPORT
  ggsave(output.tax,
    plot(tax.report),
    width = 16,
    height = 10,
    dpi = 72
  )
  
  # SAVE GEO-REPORT
  htmlwidgets::saveWidget(geo.report, file = output.geo)

  # SAVE ORGN-REPORT
  png(filename = output.orgn, width = 800, height = 400, res = 100)
    plot(orgn.report)
  dev.off()

}

```

```{r, include = FALSE}
#rmarkdown::render('palmid_dev.Rmd',
#                  output_file = output.html)
```

